File: client-libraries/javascript/src/resources/cameras.js
import Resource from './base';

/**
 * Cameras resource.
 */
class CamerasResource extends Resource {
  /**
   * List all cameras.
   * @returns {Promise<Array>} List of cameras
   */
  async list() {
    return this.get('/api/v1/cameras');
  }

  /**
   * Get camera details.
   * @param {string} cameraId - Camera ID
   * @returns {Promise<Object>} Camera details
   */
  async get(cameraId) {
    return this.get(`/api/v1/cameras/${cameraId}`);
  }

  /**
   * Create a new camera.
   * @param {Object} cameraData - Camera data
   * @returns {Promise<Object>} Created camera
   */
  async create(cameraData) {
    return this.post('/api/v1/cameras', { data: cameraData });
  }

  /**
   * Update an existing camera.
   * @param {string} cameraId - Camera ID
   * @param {Object} cameraData - Camera data
   * @returns {Promise<Object>} Updated camera
   */
  async update(cameraId, cameraData) {
    return this.put(`/api/v1/cameras/${cameraId}`, { data: cameraData });
  }

  /**
   * Delete an existing camera.
   * @param {string} cameraId - Camera ID
   * @returns {Promise<Object>} Response data
   */
  async delete(cameraId) {
    return this.delete(`/api/v1/cameras/${cameraId}`);
  }

  /**
   * Get connection status for cameras.
   * @param {string} cameraId - Camera ID (optional)
   * @returns {Promise<Object>} Connection status
   */
  async getConnectionStatus(cameraId) {
    if (cameraId) {
      return this.get(`/api/v1/cameras/${cameraId}/connection/status`);
    }
    return this.get('/api/v1/cameras/connection/status');
  }

  /**
   * Reset connection for a specific camera.
   * @param {string} cameraId - Camera ID
   * @returns {Promise<Object>} Response data
   */
  async resetConnection(cameraId) {
    return this.post(`/api/v1/cameras/${cameraId}/connection/reset`);
  }

  /**
   * Get the latest frame from a camera.
   * @param {string} cameraId - Camera ID
   * @param {Object} options - Options
   * @param {string} options.format - Image format (jpg, png, webp)
   * @param {boolean} options.bbox - Include bounding box
   * @param {boolean} options.timestamp - Include timestamp
   * @param {boolean} options.zones - Include zones
   * @param {boolean} options.mask - Include mask
   * @param {boolean} options.motion - Include motion
   * @param {boolean} options.regions - Include regions
   * @param {number} options.quality - Image quality
   * @param {number} options.height - Image height
   * @returns {Promise<Blob>} Image data
   */
  async getLatestFrame(cameraId, options = {}) {
    const { format = 'jpg', ...params } = options;
    const validFormats = ['jpg', 'jpeg', 'png', 'webp'];
    const imageFormat = validFormats.includes(format) ? format : 'jpg';

    return this.get(`/api/v1/${cameraId}/latest.${imageFormat}`, {
      params,
      headers: { Accept: 'image/*' },
      responseType: 'blob',
    });
  }

  /**
   * Get MJPEG feed URL for a camera.
   * @param {string} cameraId - Camera ID
   * @param {Object} options - Options
   * @param {number} options.fps - Frames per second
   * @param {number} options.height - Image height
   * @param {boolean} options.bbox - Include bounding box
   * @param {boolean} options.timestamp - Include timestamp
   * @param {boolean} options.zones - Include zones
   * @param {boolean} options.mask - Include mask
   * @param {boolean} options.motion - Include motion
   * @param {boolean} options.regions - Include regions
   * @returns {string} MJPEG feed URL
   */
  getMjpegFeed(cameraId, options = {}) {
    const params = new URLSearchParams();
    Object.entries(options).forEach(([key, value]) => {
      params.append(key, value);
    });
    return `${this.client.host}/api/v1/${cameraId}?${params.toString()}`;
  }

  /**
   * Get PTZ information for a camera.
   * @param {string} cameraId - Camera ID
   * @returns {Promise<Object>} PTZ information
   */
  async getPtzInfo(cameraId) {
    return this.get(`/api/v1/${cameraId}/ptz/info`);
  }

  /**
   * Control PTZ for a camera.
   * @param {string} cameraId - Camera ID
   * @param {string} command - PTZ command
   * @param {Object} options - Command options
   * @returns {Promise<Object>} Response data
   */
  async controlPtz(cameraId, command, options = {}) {
    return this.post(`/api/v1/${cameraId}/ptz/${command}`, { data: options });
  }
}

export default CamerasResource;

File: client-libraries/python/arkos_client/resources/cameras.py
"""
Cameras resource for the Arkos client library.
"""

from .base import Resource


class CamerasResource(Resource):
    """Cameras resource."""

    def list(self):
        """
        List all cameras.

        Returns:
            list: List of cameras
        """
        return self._request("GET", "/api/v1/cameras")

    def get(self, camera_id):
        """
        Get camera details.

        Args:
            camera_id (str): Camera ID

        Returns:
            dict: Camera details
        """
        return self._request("GET", f"/api/v1/cameras/{camera_id}")

    def create(self, camera_data):
        """
        Create a new camera.

        Args:
            camera_data (dict): Camera data

        Returns:
            dict: Created camera
        """
        return self._request("POST", "/api/v1/cameras", json=camera_data)

    def update(self, camera_id, camera_data):
        """
        Update an existing camera.

        Args:
            camera_id (str): Camera ID
            camera_data (dict): Camera data

        Returns:
            dict: Updated camera
        """
        return self._request("PUT", f"/api/v1/cameras/{camera_id}", json=camera_data)

    def delete(self, camera_id):
        """
        Delete an existing camera.

        Args:
            camera_id (str): Camera ID

        Returns:
            dict: Response data
        """
        return self._request("DELETE", f"/api/v1/cameras/{camera_id}")

    def get_connection_status(self, camera_id=None):
        """
        Get connection status for cameras.

        Args:
            camera_id (str, optional): Camera ID. If not provided, get status for all cameras.

        Returns:
            dict: Connection status
        """
        if camera_id:
            return self._request("GET", f"/api/v1/cameras/{camera_id}/connection/status")
        return self._request("GET", "/api/v1/cameras/connection/status")

    def reset_connection(self, camera_id):
        """
        Reset connection for a specific camera.

        Args:
            camera_id (str): Camera ID

        Returns:
            dict: Response data
        """
        return self._request("POST", f"/api/v1/cameras/{camera_id}/connection/reset")

    def get_latest_frame(self, camera_id, format="jpg", **kwargs):
        """
        Get the latest frame from a camera.

        Args:
            camera_id (str): Camera ID
            format (str, optional): Image format (jpg, png, webp). Defaults to "jpg".
            **kwargs: Additional parameters (bbox, timestamp, zones, mask, motion, regions, quality, height)

        Returns:
            bytes: Image data
        """
        valid_formats = ["jpg", "jpeg", "png", "webp"]
        if format not in valid_formats:
            format = "jpg"

        return self._request(
            "GET", f"/api/v1/{camera_id}/latest.{format}", params=kwargs, headers={"Accept": "image/*"}
        )

    def get_mjpeg_feed(self, camera_id, **kwargs):
        """
        Get MJPEG feed URL for a camera.

        Args:
            camera_id (str): Camera ID
            **kwargs: Additional parameters (fps, height, bbox, timestamp, zones, mask, motion, regions)

        Returns:
            str: MJPEG feed URL
        """
        params = "&".join([f"{k}={v}" for k, v in kwargs.items()])
        return f"{self.client.host}/api/v1/{camera_id}?{params}"

    def get_ptz_info(self, camera_id):
        """
        Get PTZ information for a camera.

        Args:
            camera_id (str): Camera ID

        Returns:
            dict: PTZ information
        """
        return self._request("GET", f"/api/v1/{camera_id}/ptz/info")
